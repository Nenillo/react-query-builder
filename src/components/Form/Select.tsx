import React from 'react';

import styled from 'styled-components';
import { colors } from '../../constants/colors';

const StyledSelect = styled.select`
  padding: 0.2rem 0.5rem;
  height: 1.6rem;
  line-height: 1.6rem;
  border: 1px solid ${colors.medium};
  border-radius: 3px;
`;

export interface SelectProps {
  values: { value: React.ReactText; label: string }[];
  selectedValue?: React.ReactText;
  emptyValue?: string;
  onChange: (value: any) => void;
  className?: string;
}

export const Select: React.FC<SelectProps> = ({
  values,
  selectedValue,
  emptyValue,
  onChange,
  className,
}) => {
  const handleChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    onChange(event.target.value);
  };

  return (
    <StyledSelect
      onChange={handleChange}
      value={selectedValue}
      className={className}
    >
      <option value="" disabled>
        {emptyValue}
      </option>
      {values.map(({ value: optionValue, label: optionLabel }) => (
        <option value={optionValue} key={optionValue}>
          {optionLabel}
        </option>
      ))}
    </StyledSelect>
  );
};
