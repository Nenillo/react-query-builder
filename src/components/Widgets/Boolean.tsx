import React, { useContext } from 'react';
import { BuilderContext } from '../Context'
import { clone } from '../../utils/clone';

export interface BooleanProps {
  selectedValue: boolean;
  id: string;
}

export const Boolean: React.FC<BooleanProps> = ({ selectedValue, id }) => {
  const { data, setData, onChange, components } = useContext(BuilderContext);

  const { form } = components;

  const handleChange = (value: boolean) => {
    const clonedData = clone(data);
    const parentIndex = clonedData.findIndex((item: any) => item.id === id);

    clonedData[parentIndex].value = value;

    setData(clonedData);
    onChange(clonedData);
  };

  if (form) {
    return <form.Switch onChange={handleChange} switched={selectedValue} />;
  }

  return null;
};
