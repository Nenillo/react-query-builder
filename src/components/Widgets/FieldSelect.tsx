import React, { useContext } from 'react';
import { BuilderContext } from '../Builder';
import { clone } from '../../utils/clone';
import { isOptionList } from '../../utils/types';

interface FieldSelectProps {
  selectedValue: string;
  id: string;
}

export const FieldSelect: React.FC<FieldSelectProps> = ({
  selectedValue,
  id,
}) => {
  const { fields, data, setData, onChange, components } = useContext(
    BuilderContext
  );

  const {
    form: { Select },
  } = components;

  const handleChange = (value: React.ReactText) => {
    const clonedData = clone(data);
    const parentIndex = clonedData.findIndex((item: any) => item.id === id);
    const nextField = fields.filter(item => item.field === value)[0];

    clonedData[parentIndex].field = value;
    delete clonedData[parentIndex].value;
    delete clonedData[parentIndex].operators;
    delete clonedData[parentIndex].operator;

    switch (nextField.type) {
      case 'BOOLEAN':
        clonedData[parentIndex].value = false;
        break;

      case 'DATE':
        if (
          nextField.operators &&
          ['BETWEEN', 'NOT_BETWEEN'].includes(nextField.operators[0])
        ) {
          clonedData[parentIndex].value = ['', ''];
        } else {
          clonedData[parentIndex].value = '';
        }

        clonedData[parentIndex].operator =
          nextField.operators && nextField.operators[0];
        clonedData[parentIndex].operators = nextField.operators;
        break;

      case 'TEXT':
        if (
          nextField.operators &&
          ['BETWEEN', 'NOT_BETWEEN'].includes(nextField.operators[0])
        ) {
          clonedData[parentIndex].value = ['', ''];
        } else {
          clonedData[parentIndex].value = '';
        }

        clonedData[parentIndex].operator =
          nextField.operators && nextField.operators[0];
        clonedData[parentIndex].operators = nextField.operators;
        break;

      case 'NUMBER':
        if (
          nextField.operators &&
          ['BETWEEN', 'NOT_BETWEEN'].includes(nextField.operators[0])
        ) {
          clonedData[parentIndex].value = [0, 0];
        } else {
          clonedData[parentIndex].value = 0;
        }

        clonedData[parentIndex].operator =
          nextField.operators && nextField.operators[0];
        clonedData[parentIndex].operators = nextField.operators;
        break;

      case 'LIST':
        if (isOptionList(nextField.value)) {
          clonedData[parentIndex].value = nextField.value[0].value;
        }

        clonedData[parentIndex].operator =
          nextField.operators && nextField.operators[0];
        clonedData[parentIndex].operators = nextField.operators;
        break;

      case 'MULTI_LIST':
        if (isOptionList(nextField.value)) {
          clonedData[parentIndex].value = [];
        }

        clonedData[parentIndex].operator =
          nextField.operators && nextField.operators[0];
        clonedData[parentIndex].operators = nextField.operators;
        break;
      case 'STATEMENT':
        clonedData[parentIndex].value = nextField.value;
        break;
    }

    setData(clonedData);
    onChange(clonedData);
  };

  const fieldNames = fields.map(field => ({
    value: field.field,
    label: field.label,
  }));

  return (
    <Select
      values={fieldNames}
      selectedValue={selectedValue}
      onChange={handleChange}
    />
  );
};
