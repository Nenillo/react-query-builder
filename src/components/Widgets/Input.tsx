import React, { useContext } from 'react';
import { isReactTextArray, isUndefined } from '../../utils/types';
import { BuilderContext } from '../Builder';
import { clone } from '../../utils/clone';

interface InputProps {
  type: 'date' | 'number' | 'text';
  value: React.ReactText | React.ReactText[];
  id: string;
}

export const Input: React.FC<InputProps> = ({ type, value, id }) => {
  const { data, setData, onChange, components } = useContext(BuilderContext);

  const {
    form: { Input: InputBase },
  } = components;

  const handleChange = (value: any, index?: number) => {
    const clonedData = clone(data);
    const parentIndex = clonedData.findIndex((item: any) => item.id === id);

    if (!isUndefined(index)) {
      clonedData[parentIndex].value[index] = value;
    } else {
      clonedData[parentIndex].value = value;
    }

    setData(clonedData);
    onChange(clonedData);
  };

  if (isReactTextArray(value)) {
    return (
      <>
        <InputBase
          type={type}
          value={value[0]}
          onChange={(value: any) => handleChange(value, 0)}
        />
        <InputBase
          type={type}
          value={value[1]}
          onChange={(value: any) => handleChange(value, 1)}
        />
      </>
    );
  } else {
    return <InputBase type={type} value={value} onChange={handleChange} />;
  }
};
