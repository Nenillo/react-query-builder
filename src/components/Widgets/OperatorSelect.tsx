import React, { useContext } from 'react';
import { BuilderContext, BuilderFieldOperator } from '../Builder';
import { clone } from '../../utils/clone';
import { isReactTextArray } from '../../utils/types';

interface OperatorSelectProps {
  values: { value: BuilderFieldOperator; label?: string }[];
  selectedValue?: BuilderFieldOperator;
  id: string;
}

export const OperatorSelect: React.FC<OperatorSelectProps> = ({
  values,
  selectedValue,
  id,
}) => {
  const { fields, data, setData, onChange, components, strings } = useContext(
    BuilderContext
  );

  const { form } = components;

  const handleChange = (value: BuilderFieldOperator) => {
    const clonedData = clone(data);
    const parentIndex = clonedData.findIndex((item: any) => item.id === id);
    const fieldIndex = fields.findIndex(
      (item: any) => clonedData[parentIndex].field === item.field
    );

    if (['DATE', 'TEXT', 'NUMBER'].includes(fields[fieldIndex].type)) {
      if (
        !['BETWEEN', 'NOT_BETWEEN'].includes(value) &&
        isReactTextArray(clonedData[parentIndex].value)
      ) {
        clonedData[parentIndex].value = '';
      } else if (
        ['BETWEEN', 'NOT_BETWEEN'].includes(value) &&
        !isReactTextArray(clonedData[parentIndex].value)
      ) {
        clonedData[parentIndex].value = ['', ''];
      }
    }

    clonedData[parentIndex].operator = value;

    setData(clonedData);
    onChange(clonedData);
  };

  if (form && strings.form) {
    return (
      <form.Select
        values={values}
        selectedValue={selectedValue}
        emptyValue={strings.form.selectYourValue}
        onChange={handleChange}
      />
    );
  }

  return null;
};
