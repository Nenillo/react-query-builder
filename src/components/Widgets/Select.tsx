import React, { useContext } from 'react';
import { BuilderContext } from '../Builder';
import { clone } from '../../utils/clone';

export interface SelectProps {
  selectedValue: React.ReactText;
  values: { value: React.ReactText; label: string }[];
  id: string;
}

export const Select: React.FC<SelectProps> = ({
  selectedValue,
  values,
  id,
}) => {
  const { data, setData, onChange, components } = useContext(BuilderContext);

  const {
    form: { Select: SelectBase },
  } = components;

  const handleChange = (value: React.ReactText) => {
    const clonedData = clone(data);
    const parentIndex = clonedData.findIndex((item: any) => item.id === id);

    clonedData[parentIndex].value = value;

    setData(clonedData);
    onChange(clonedData);
  };

  return (
    <SelectBase
      onChange={handleChange}
      selectedValue={selectedValue}
      values={values}
    />
  );
};
