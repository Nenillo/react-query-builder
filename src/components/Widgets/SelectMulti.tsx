import React, { useContext } from 'react';
import { BuilderContext } from '../Builder';
import { clone } from '../../utils/clone';

export interface SelectMultiProps {
  values: { value: string; label: string }[];
  selectedValue: string[];
  id: string;
}

export const SelectMulti: React.FC<SelectMultiProps> = ({
  values,
  selectedValue,
  id,
}) => {
  const { data, setData, onChange, components, strings } = useContext(
    BuilderContext
  );

  const { form } = components;

  const handleChange = (value: string) => {
    if (setData && onChange) {
      const clonedData = clone(data);
      const parentIndex = clonedData.findIndex((item: any) => item.id === id);

      clonedData[parentIndex].value = clonedData[parentIndex].value.filter(
        (item: any) => item !== value
      );
      clonedData[parentIndex].value.push(value);

      setData(clonedData);
      onChange(clonedData);
    }
  };

  const handleDelete = (value: string) => {
    const clonedData = clone(data);
    const parentIndex = clonedData.findIndex((item: any) => item.id === id);

    clonedData[parentIndex].value = clonedData[parentIndex].value.filter(
      (item: any) => item !== value
    );

    setData(clonedData);
    onChange(clonedData);
  };

  if (form && strings.form) {
    return (
      <form.SelectMulti
        onChange={handleChange}
        onDelete={handleDelete}
        selectedValue={selectedValue}
        emptyValue={strings.form.selectYourValue}
        values={values}
      />
    );
  }

  return null;
};
