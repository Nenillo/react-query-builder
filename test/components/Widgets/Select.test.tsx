import { Select } from '../../../src/components/Widgets/Select';
import {
  BuilderContext,
  BuilderFieldProps,
  BuilderComponentsProps,
  defaultComponents,
} from '../../../src/components/Builder';
import { shallow, mount } from 'enzyme';
import React from 'react';

const components: BuilderComponentsProps = defaultComponents;
const fields: BuilderFieldProps[] = [
  {
    field: 'MOCK_FIELD',
    label: 'Mock Field',
    type: 'LIST',
    value: [
      {
        value: 'test',
        label: 'test',
      },
    ],
  },
];
const data: any[] = [
  {
    id: 'test',
    field: 'MOCK_FIELD',
    value: 'test',
  },
];
const selectValues = [
  {
    value: 'test',
    label: 'test',
  },
];
const strings = { form: {} };
const setData = jest.fn();
const onChange = () => jest.fn();

describe('#components/Widgets/Select', () => {
  it('Tests Snapshot', () => {
    expect(
      shallow(
        <BuilderContext.Provider
          value={{ components, fields, data, strings, setData, onChange }}
        >
          <Select id="test" selectedValue={''} values={selectValues} />
        </BuilderContext.Provider>
      )
    ).toMatchSnapshot();
  });

  it('Tests user interaction', () => {
    const wrapper = mount(
      <BuilderContext.Provider
        value={{ components, fields, data, strings, setData, onChange }}
      >
        <Select id="test" selectedValue={''} values={selectValues} />
      </BuilderContext.Provider>
    );

    const select = wrapper.find('select').first();
    select.simulate('change', { target: { value: 'test' } });
  });

  it('Test no form components scenario', () => {
    const wrapper = mount(
      <BuilderContext.Provider
        value={{ components: {}, fields, data, strings, setData, onChange }}
      >
        <Select id="test" selectedValue={''} values={selectValues} />
      </BuilderContext.Provider>
    );

    expect(Object.keys(wrapper).length).toBe(0);
  });
});
