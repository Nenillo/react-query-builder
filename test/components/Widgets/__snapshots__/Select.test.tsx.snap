// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#components/Widgets/Select Tests Snapshot 1`] = `
<Component
  id="test"
  selectedValue=""
  values={
    Array [
      Object {
        "label": "test",
        "value": "test",
      },
    ]
  }
/>
`;
